src/pool/srv_pool.c

```diff
@@ -5464,6 +5473,19 @@ pool_svc_cancel_and_wait_reconf(struct pool_svc *svc)
 	reconf_cancel_and_wait(&svc->ps_reconf);
 }
 
+/**
+ * Perform PS reconfigurations (if necessary) synchronously. This is currently
+ * for the chk module only.
+ */
+void
+ds_pool_svc_reconf(struct ds_pool_svc *svc)
+{
+	struct pool_svc *s = pool_ds2svc(svc);
+
+	pool_svc_schedule_reconf(s, true /* for_chk */);
+	reconf_wait(&s->ps_reconf);
```

Comment:
Refamiliarizing with the code after reviewing earlier and having some older discussions about cat_recovery I don't recall well. Apologies if it's a silly question: In this scenario (for_chk == true), will it be the case that there is no pool_svc_reconf_ult running? i.e., the reconf_cancel_and_wait() call in pool_svc_schedule_reconf() will only lock and unlock the mutex and not actually request to cancel an in progress ULT? Just wondering otherwise if there would be a race to be unblocked on the CV wait between this reconf_wait() call and the reconf_cancel_and_wait() call.

Reply:
Currently there should be no PS reconf ULTs when the chk module calls this function. On feature/cat_recovery I've added engine_in_check calls to pool_svc_schedule_reconf so that the PS won't schedule any automatic reconf ULTs in the "check" mode, which is set to true when the daos_engine process starts.

Sorry, I don't get the problem with the "race" if there were existing PS reconf ULTs. Could you elaborate a little, @kccain?

Reply:
I was confused. In check mode, even if a PS reconf ULT were running (let's call it ULT execution A) at the time of this pool_svc_reschedule_reconf(), then the reconf_cancel_and_wait() call would cause that ULT to be cancelled, then would cause another ULT (let's call it ULT execution B) to occur. Then the reconf_wait() call here would wait for execution B to be done.

My confusion was if somehow we could have 2 threads waiting on reconf->psc_cv and a cancelled ULT calls cond_broadcast then we don't know which of the 2 threads waiting on psc_cv would be awoken first.

Looks like in your latest commit though this reconf_wait() call is removed.

Reply:
There will be at most one PS reconf ULT at any moment during the term of a PS leader. We implement that by always cancel any existing PS reconf ULT and wait for it to terminate first (the reconf_cancel_and_wait call) in pool_svc_schedule_reconf. A new PS reconf ULT will take care of what any canceled one should do.

In this very chk scenario, besides the absence of existing PS reconf ULTs, I think it suffices to only trigger a best-effort reconf.